src/rio.c

```diff
@@ -108,19 +109,49 @@ void rioInitWithBuffer(rio *r, sds s) {
 
 /* Returns 1 or 0 for success/failure. */
 static size_t rioFileWrite(rio *r, const void *buf, size_t len) {
-    size_t retval;
+    if (!r->io.file.autosync) return fwrite(buf,len,1,r->io.file.fp);
 
-    retval = fwrite(buf,len,1,r->io.file.fp);
-    r->io.file.buffered += len;
+    size_t nwritten = 0;
+    while (len != nwritten) {
+        assert(r->io.file.autosync > r->io.file.buffered);
+        size_t nalign = (size_t)(r->io.file.autosync - r->io.file.buffered);
+        size_t towrite = nalign > len-nwritten ? len-nwritten : nalign;
+
+        if (fwrite((char*)buf+nwritten,towrite,1,r->io.file.fp) == 0) return 0;
+        nwritten += towrite;
+        r->io.file.buffered += towrite;
+
+        if (r->io.file.buffered >= r->io.file.autosync) {
+            fflush(r->io.file.fp);
+
+            size_t processed = r->processed_bytes + nwritten;
+            assert(processed % r->io.file.autosync == 0);
+            assert(r->io.file.buffered == r->io.file.autosync);
```

Comment:
i think we can improve the loop above, and these bunch of asserts are good to demonstrate what the loop aims for.
a small 10 byte write, could happen to fall on the boundary and be written in two fwrite calls.

instead, maybe we can:
1. make sure we never write a single chunk of more than 4mb (so if we have a big input buffer we write it in parts in a loop), and:
2. separately, we track when was the last time we calls fsync, and call it if the unsynced part is more than 4mb.
3. lastly, if we sync_file_range, we just avoid waiting for the last 4mb (see below)